optuna/study/_optimize.py

```diff
@@ -194,15 +192,9 @@ def _run_trial(
     value_or_values: Optional[Union[float, Sequence[float]]] = None
     func_err: Optional[Union[Exception, KeyboardInterrupt]] = None
     func_err_fail_exc_info: Optional[Any] = None
-    stop_event: Optional[Event] = None
-    thread: Optional[Thread] = None
 
-    if is_heartbeat_enabled(study._storage):
-        assert isinstance(study._storage, BaseHeartbeat)
-        heartbeat = study._storage
-        stop_event = Event()
-        thread = Thread(target=_record_heartbeat, args=(trial._trial_id, heartbeat, stop_event))
-        thread.start()
+    heartbeat_thread = get_heartbeat_thread(study._storage)
+    heartbeat_thread.start(trial._trial_id)
```

Comment:
[Question] Can `get_heartbeat_thread` create a context-manager?

I guess it enable us to use the `with` statement to ensure that `heartbeat_thread.join` is always called. I prototyped it and it seemed to work as far as I checked in unit tests (i.e., `tests/study_tests`).

https://github.com/g-votte/optuna/compare/refactor-heartbeat-thread...toshihikoyanase:heatbeat-context-manager?expand=1

Reply:
Cool. I agree that supporting the `with` scope makes the resource management easier and safer.

I committed a change based on your proposal: https://github.com/optuna/optuna/pull/3609/commits/148993accf07a36cf4d90b921b051477d03757e2

There're some additional changes to remove `trial_id` from the `start` arguments, which make `__enter__` and `start` more symmetry.